ambry-router/src/test/java/com/github/ambry/router/ParanoidDurabilityOperationTrackerTest.java

```diff
@@ -0,0 +1,267 @@
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.network.Port;
+import com.github.ambry.network.PortType;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+import org.junit.Assert;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+
+/**
+ * Unit tests for {@link ParanoidDurabilityOperationTracker}.
+ *
+ * {@link ParanoidDurabilityOperationTracker} needs to control writes to both local and remote replicas, and ensure that
+ * the operation is successful only if the required number of replicas in both local and remote data centers are successful.
+ */
+//@RunWith(Parameterized.class)
+public class ParanoidDurabilityOperationTrackerTest {
+  private static final int PORT = 7777;
+  private MockPartitionId mockPartition;
+  private String localDcName;
+  private MockClusterMap mockClusterMap;
+  private final LinkedList<ReplicaId> inflightReplicas = new LinkedList<>();
+  private final Set<ReplicaId> repetitionTracker = new HashSet<>();
+
+
+  public ParanoidDurabilityOperationTrackerTest() { }
+
+  /**
+   * A vanilla test with good replicas everywhere and a successful outcome.
+   */
+  @Test
+  public void basicSuccessTest() {
+    localDcName = new String("dc-0");
+    List<Port> portList = Collections.singletonList(new Port(PORT, PortType.PLAINTEXT));
+    List<String> mountPaths = Collections.singletonList("mockMountPath");
+    List<ReplicaSpec> testReplicas = new ArrayList<ReplicaSpec>(Arrays.asList(new ReplicaSpec[] {
+        new ReplicaSpec(localDcName, ReplicaState.LEADER, portList, mountPaths),
+        new ReplicaSpec(localDcName, ReplicaState.STANDBY, portList, mountPaths),
+        new ReplicaSpec(localDcName, ReplicaState.STANDBY, portList, mountPaths),
+        new ReplicaSpec("remote-1", ReplicaState.LEADER, portList, mountPaths),
+        new ReplicaSpec("remote-1", ReplicaState.STANDBY, portList, mountPaths),
+        new ReplicaSpec("remote-1", ReplicaState.STANDBY, portList, mountPaths),
+        new ReplicaSpec("remote-2", ReplicaState.LEADER, portList, mountPaths),
+        new ReplicaSpec("remote-2", ReplicaState.STANDBY, portList, mountPaths),
+        new ReplicaSpec("remote-2", ReplicaState.STANDBY, portList, mountPaths) }));
+
+    mockPartition = new MockPartitionId();
+    mockClusterMap = new MockClusterMap(false, buildDataNodeList(testReplicas), 1, Collections.singletonList(mockPartition), localDcName);
+    populateReplicaList(testReplicas);
+
+    ParanoidDurabilityOperationTracker ot = getParanoidDurabilityOperationTracker(2, 1, 2, 1, true);
+    while (!ot.hasSucceeded()) {
+      sendRequests(ot, 3);
+      for (int i = 0; i < 3; i++) {
+        ot.onResponse(inflightReplicas.poll(), TrackedRequestFinalState.SUCCESS);
+      }
```

Comment:
the parallelism for local is 2 and remote is 1, we are expecting to send 3 requests out, which is right. Can we verify that the first two inflight replicas are local and the last one is remote. And after all replicas return success, the operation tracker should return success.

Reply:
OK, I added the verification to count local and remote successes. I think it was already checking that the operation tracker was returning success, but also added a check to confirm that no failure was detected in successful cases.